source/slang/diagnostics.h

```diff
@@ -217,6 +218,32 @@ namespace Slang
         void diagnoseRaw(
             Severity    severity,
             const UnownedStringSlice& message);
+
+            /// During propagation of an exception for an internal
+            /// error, note that this source location was involved
+        void noteInternalErrorLoc(SourceLoc const& loc);
+    };
+
+        /// An `ISlangWriter` that writes directly to a diagnostic sink.
+    class DiagnosticSinkWriter : public AppendBufferWriter
+    {
+    public:
+        typedef AppendBufferWriter Super;
+
+        DiagnosticSinkWriter(DiagnosticSink* sink)
+            : Super(WriterFlag::IsStatic)
+            , m_sink(sink)
+        {}
+
+        // ISlangWriter
+        SLANG_NO_THROW virtual SlangResult SLANG_MCALL write(const char* chars, size_t numChars) SLANG_OVERRIDE
```

Comment:
I completely agree that the abstraction layers are being inverted here. This writer type isn’t something I’d actually advocate using more, but if was the quickest way to get the IR dumping logic to write to where it should have in the first place, so we can do a real fix later.

The main alternative was to write into a temporary string/buffer and then write that to the sink, which would maybe be cleaner, but I figured it would also add overhead. We can clean this up in whatever way we like down the road.